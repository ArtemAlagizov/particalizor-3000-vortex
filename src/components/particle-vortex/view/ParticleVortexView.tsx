import React from "react";
import { IParticleVortexView } from "./IParticleVortexView";
import { ParticleVortex } from "particalizor-3000";
import { Styled } from "../styled";
import ContainerDimensions from "react-container-dimensions";

const ParticleVortexShowcaseView: React.FC<IParticleVortexView> = ({
  particleNumberValue,
  particleTraceWidthValue,
  particleLifeTimeValue,
  backgroundColor,
  vortexNumberValue,
  onShowCaseItemClick,
  showCases
}) => {
  return (
    <Styled.ShowcaseComponentContainer>
      <Styled.MenuItemsContainer>
        {showCases.map(showCase => (
          <Styled.MenuItem
            key={showCase.id}
            onClick={() => onShowCaseItemClick(showCase.id)}
          >
            <div> {showCase.title} </div>
          </Styled.MenuItem>
        ))}
      </Styled.MenuItemsContainer>

      <ContainerDimensions>
        {({ height, width }) => (
          <ParticleVortex
            imageWidth={width}
            imageHeight={height}
            vortexNumber={vortexNumberValue}
            particleTraceWidth={particleTraceWidthValue}
            particleNumber={particleNumberValue}
            particleLifeTime={particleLifeTimeValue}
            backgroundColor={backgroundColor}
          />
        )}
      </ContainerDimensions>
    </Styled.ShowcaseComponentContainer>
  );
};

export default ParticleVortexShowcaseView;
