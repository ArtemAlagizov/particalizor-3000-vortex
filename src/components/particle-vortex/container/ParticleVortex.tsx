import React, { useState } from "react";
import ParticleVortexView from "../view/ParticleVortexView";
import { NumberParam, StringParam, useQueryParam } from "use-query-params";
import { isNil } from "ramda";
import { useHistory } from "react-router";
import { IShowCase } from "./IShowCase";

const ParticleVortex: React.FC = () => {
  const [isMenuVisible, setIsMenuVisible] = useState<boolean>(false);

  const [vortexNumber] = useQueryParam("vortexNumber", NumberParam);
  const [particleNumber] = useQueryParam("particleNumber", NumberParam);
  const [particleTraceWidth] = useQueryParam("particleTraceWidth", NumberParam);
  const [particleLifeTime] = useQueryParam("particleLifeTime", NumberParam);
  const [backgroundColor] = useQueryParam("backgroundColor", StringParam);

  const selectedParticleNumber = isNil(particleNumber) ? 30 : particleNumber;
  const selectedParticleTraceWidth = isNil(particleTraceWidth)
    ? 600
    : particleTraceWidth;
  const selectedParticleLifeTime = isNil(particleLifeTime)
    ? 1100
    : particleLifeTime;
  const selectedVortexNumber = isNil(vortexNumber) ? 2 : vortexNumber;
  const selectedBackgroundColor = isNil(backgroundColor)
    ? "#33344c"
    : backgroundColor;

  const showCases: IShowCase[] = [
    {
      id: "flowMachine",
      urlPath:
        "?particleLifeTime=30&particleTraceWidth=40&vortexNumber=3&particleNumber=400&backgroundColor=%231a414e",
      title: "FLOW MACHINE"
    },
    {
      id: "dimensionalSpin",
      urlPath:
        "?particleLifeTime=100&particleTraceWidth=250&vortexNumber=4&particleNumber=20&backgroundColor=%232b2b59",
      title: "DIMENSIONAL SPIN"
    },
    {
      id: "snakeUniverse",
      urlPath:
        "?particleLifeTime=1000&particleTraceWidth=30&vortexNumber=3&particleNumber=200&backgroundColor=%231b293b",
      title: "SNAKE UNIVERSE"
    },
    {
      id: "galacticJam",
      urlPath:
        "?particleLifeTime=1000&particleTraceWidth=1&vortexNumber=3&particleNumber=300&backgroundColor=%2315151a",
      title: "GALACTIC JAM"
    },
    {
      id: "windmillOfTime",
      urlPath: "",
      title: "WINDMILL OF TIME"
    }
  ];

  const history = useHistory();

  const onMenuToggle = () => setIsMenuVisible(!isMenuVisible);
  const onGithubLinkClick = () =>
    (window.location.href = "https://github.com/ArtemAlagizov/relaxator");
  const onShowCaseItemClick = (showCaseId: string) => {
    const selectedShowCase = showCases.find(
      showCase => showCase.id === showCaseId
    );
    const redirectPath = isNil(selectedShowCase)
      ? ""
      : selectedShowCase.urlPath;

    history.push(redirectPath);
  };

  return (
    <ParticleVortexView
      particleLifeTimeValue={selectedParticleLifeTime}
      particleNumberValue={selectedParticleNumber}
      particleTraceWidthValue={selectedParticleTraceWidth}
      vortexNumberValue={selectedVortexNumber}
      backgroundColor={selectedBackgroundColor}
      onShowCaseItemClick={onShowCaseItemClick}
      showCases={showCases}
      isMenuVisible={isMenuVisible}
      onMenuToggle={onMenuToggle}
      onGithubLinkClick={onGithubLinkClick}
    />
  );
};

export default ParticleVortex;
