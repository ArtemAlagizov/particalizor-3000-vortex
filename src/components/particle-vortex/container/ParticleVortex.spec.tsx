import React from "react";
import { mount, shallow } from "enzyme";
import ParticleVortex from "./ParticleVortex";
import ParticleVortexView from "../view/ParticleVortexView";

jest.mock("react-router-dom");
jest.mock("react-router", () => ({
  useHistory: () => ({
    push: jest.fn()
  })
}));

describe("Given a ParticleVortex", () => {
  describe("when rendering", () => {
    it("should match snapshot", () => {
      const component = shallow(<ParticleVortex />);

      expect(component).toMatchSnapshot();
    });
    it("should render ParticleVortexView", () => {
      const component = mount(<ParticleVortex />);

      expect(component.find(ParticleVortexView).length).toBe(1);
    });
  });
});
