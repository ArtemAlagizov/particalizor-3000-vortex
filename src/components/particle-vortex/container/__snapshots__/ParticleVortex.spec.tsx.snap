// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a ParticleVortex when rendering should match snapshot 1`] = `
<ParticleVortexShowcaseView
  backgroundColor="#33344c"
  isMenuVisible={false}
  onGithubLinkClick={[Function]}
  onMenuToggle={[Function]}
  onShowCaseItemClick={[Function]}
  particleLifeTimeValue={1100}
  particleNumberValue={30}
  particleTraceWidthValue={600}
  showCases={
    Array [
      Object {
        "id": "flowMachine",
        "title": "FLOW MACHINE",
        "urlPath": "?particleLifeTime=30&particleTraceWidth=40&vortexNumber=3&particleNumber=400&backgroundColor=%231a414e",
      },
      Object {
        "id": "dimensionalSpin",
        "title": "DIMENSIONAL SPIN",
        "urlPath": "?particleLifeTime=100&particleTraceWidth=250&vortexNumber=4&particleNumber=20&backgroundColor=%232b2b59",
      },
      Object {
        "id": "snakeUniverse",
        "title": "SNAKE UNIVERSE",
        "urlPath": "?particleLifeTime=1000&particleTraceWidth=30&vortexNumber=3&particleNumber=200&backgroundColor=%231b293b",
      },
      Object {
        "id": "galacticJam",
        "title": "GALACTIC JAM",
        "urlPath": "?particleLifeTime=1000&particleTraceWidth=0.7&vortexNumber=3&particleNumber=200&backgroundColor=%2315151a",
      },
      Object {
        "id": "timeWindmill",
        "title": "TIME WINDMILL",
        "urlPath": "",
      },
      Object {
        "id": "motionOcean",
        "title": "MOTION OCEAN",
        "urlPath": "?particleLifeTime=100&particleTraceWidth=250&vortexNumber=4&particleNumber=1220&backgroundColor=%232b2b59",
      },
      Object {
        "id": "wave13",
        "title": "WAVE 13",
        "urlPath": "?particleLifeTime=1000&particleTraceWidth=150&vortexNumber=4&particleNumber=13&backgroundColor=%232b2b59",
      },
    ]
  }
  vortexNumberValue={2}
/>
`;
